src/Microsoft.AspNet.OData.Shared/Formatter/Serialization/ODataResourceSerializer.cs

```diff
@@ -142,22 +142,30 @@ private void WriteDeltaResource(object graph, ODataWriter writer, ODataSerialize
             }
         }
 
-        private void WriteDeltaComplexProperties(IEnumerable<IEdmStructuralProperty> complexProperties,
+        private void WriteDeltaComplexProperties(SelectExpandNode selectExpandNode,
             ResourceContext resourceContext, ODataWriter writer)
         {
-            Contract.Assert(complexProperties != null);
             Contract.Assert(resourceContext != null);
             Contract.Assert(writer != null);
 
+            if (selectExpandNode.SelectedComplexes == null)
+            {
+                return;
+            }
+            IEnumerable<IEdmStructuralProperty> complexProperties = selectExpandNode.SelectedComplexes.Keys;
+
             if (null != resourceContext.EdmObject && resourceContext.EdmObject.IsDeltaResource())
             {
                 IDelta deltaObject = resourceContext.EdmObject as IDelta;
                 IEnumerable<string> changedProperties = deltaObject.GetChangedPropertyNames();
+
                 complexProperties = complexProperties.Where(p => changedProperties.Contains(p.Name));
             }
 
-            foreach (IEdmStructuralProperty complexProperty in complexProperties)
+            foreach (var selectedComplex in complexProperties)
             {
+                IEdmStructuralProperty complexProperty = selectedComplex;
```

Comment:
Why type as var in the foreach and then cast? Why not just:
foreach (IEdmStructuralProperty complexProperty in complexProperties) and remove 167?

Reply:
Oh, yes. Forgot to change this when i addressed the comment from Kanish. Will update it.

---
In reply to: [354992300](https://github.com/OData/WebApi/pull/1941#discussion_r354992300) [](ancestors = 354992300)